### PR TITLE
Refactoring dynamic loss to compose any loss method, not just MSE.

### PR DESCRIPTION
**Pull request by alxmrs**
We now use a builder function to create the loss function. Loss can be defined as a simple metric, or a more sophisticated loss mixed in with a metric, such as dynamic loss. This should allow us to combine MAE with dynamic loss, for example.

My hope with this change is to follow up with experimenting with this idea from the Cambdige/NVIDIA paper: https://ar5iv.labs.arxiv.org/html/2410.07472v1#S4.SS10. 

(TL;DR: I have reason to believe that dynamic MAE will perform better than dynamic MSE).

### CODE REVIEW

File: src/ocean_emulators/config.py [hunk 1]
```diff
@@ -670,6 +673,7 @@ class TrainConfig(TopLevelConfig):
     learning_rate: float = 2e-4
     scheduler: SchedulerConfig | None = None
     loss: LossType = "mse"
+    loss_extension: LossExtensionType = "off"
```
Comment (jder) [inline src/ocean_emulators/config.py]: It seems confusing to me that this makes 2 ways to ask for the dynamic loss now. I think we should either totally separate this from loss (ie remove `mse_dynamic` and break backwards compatibility) or remove `loss_extension` (and just use `loss: mae_dynamic`). 

🐑 If you keep it, don't like the name "extension" -- it seems pretty generic. If you want to separate the metric from the scaling, how about `loss_scaling`? 

Alternatively we could go all out and have a LossConfig class so we can support eg:

```yaml
loss:
  type: dynamic
  limit: false
  metric: mae
```

and do something like `Location`'s `string_to_unresolved` so that `loss: mae` still works. This might be worthwhile if @amogh-gulati's PR about ensemble losses will want a more configurable set of loss functions here but not sure if that's needed or not.

**Review by jder — CHANGES_REQUESTED** [review]

File: src/ocean_emulators/config.py (hunk 1 above)
Comment (amogh-gulati) [reply src/ocean_emulators/config.py]: It will be helpful to have dynamic loss, at least for CRPS!

File: src/ocean_emulators/config.py [hunk 2]
```diff
@@ -648,14 +666,110 @@ def build(self, output_dir: Path, device: torch.device) -> Profiler:
...
-    "mse_mae",
-    "mse_dynamic",
-    "mse_dynamic_no_limit",
+]
+LossType = Literal["standard", "dynamic"]
+
+
+class LossConfig(pydantic.BaseModel):
+    metric: LossMetric = "mse"
+    type: LossType = "standard"
```
Comment (jder) [inline src/ocean_emulators/config.py]: My goal in suggesting a separate class and "type" member was to permit different types with different members. For example, the "dynamic" type would have a `limit` field, the "standard" type would not. (Like S3Location and LocalLocation)

This would then let us have eg a "gradient" type which includes how much to weigh between the gradients and the underlying metric loss without having a "gradient_weight" member here which does nothing most of the time.

File: src/ocean_emulators/config.py [hunk 3]
```diff
@@ -648,14 +666,110 @@ def build(self, output_dir: Path, device: torch.device) -> Profiler:
...
+            f"  loss:\n"
+            f"    metric: mse\n"
+            f"    type: dynamic\n"
+            f"    limit: false"
+        )
+    return candidate
+
+
+Loss = Annotated[
+    Annotated[LossConfig, WithJsonSchema({"type": "string"})],
```
Comment (jder) [inline src/ocean_emulators/config.py]: I believe this will cause the JSON schema to reject the structured form (ie you will get red squiggles in the editor). Depending on how you resolve the above comment, you probably want either this to be `|`'d with another type (as it is in Location) or else you'll need to use a [fancier method](https://docs.pydantic.dev/latest/concepts/json_schema/#implementing-__get_pydantic_json_schema__) to extend rather than totally replacing the JSON schema.

**Review by jder — COMMENTED** [review]
I think maybe I wasn't very clear about my suggestion. Happy to discuss live or else if this is turning into too much of a diversion I think it's fine to go back to the previous strategy where we have `loss` of `mse` or `mae` or whatever and a `loss_scaling` (or whatever name) of `none` or `dynamic`.

File: src/ocean_emulators/train.py [hunk 4]
```diff
@@ -232,59 +222,21 @@ def __init__(self, cfg: TrainConfig) -> None:
...
+        stds = self.src.stds[self.prognostic_var_names]
+        scaling_residuals = None
+        if self.data_container.scaling_residuals is not None:
+            scaling_residuals = self.data_container.scaling_residuals[
+                self.prognostic_var_names
+            ]
+        self.loss_fn: LossFn = cfg.loss.build(
+            hist=cfg.data.hist,
+            wet=self.wet,
+            y_coord=self.data.y,
```
Comment (jder) [inline src/ocean_emulators/train.py]: (You might want to just delete these losses we have never tried rather than trying to fix whatever is going wrong here)

File: src/ocean_emulators/config.py [hunk 5]
```diff
@@ -648,17 +659,69 @@ def build(self, output_dir: Path, device: torch.device) -> Profiler:
...
+        case "mse_diff_weighted":
+            loss_fn = partial(decomposed_mse_diff_weighted, wet=wet)
+        case "mse_cos_weighted":
+            area_weights = np.sqrt(np.cos(np.deg2rad(y_coord))).to_numpy()
+            area_weights = torch.from_numpy(area_weights).to(device=device)
+            loss_fn = partial(decomposed_mse_cos_weighted, wet=wet, cos=area_weights)
+        case _:
+            assert_never(metric)
+
+    if isinstance(loss_cfg, DynamicLossConfig):
```
Comment (alxmrs) [inline src/ocean_emulators/config.py]: I am on the fence if this should delegate a `build` method on top of `DynamicLossConfig`.

File: src/ocean_emulators/utils/loss.py [hunk 6]
```diff
@@ -82,23 +93,25 @@ class MseDynamic:
     See: https://openathena.slack.com/archives/C08CYM42DT3/p1752275713570969
     """
 
-    N_WINDOW = 25
     """Rolling window size to average over. (~number of steps)"""
+    N_WINDOW = 25
```
Comment (jder) [inline src/ocean_emulators/utils/loss.py]: As I said [last time you asked about this](https://github.com/Open-Athena/Ocean_Emulator/pull/310#discussion_r2252375402), I believe the original order is the standard. Do you have a reference for this order?

File: src/ocean_emulators/utils/loss.py [hunk 7]
```diff
@@ -82,23 +93,25 @@ class MseDynamic:
...
 
     def __init__(
         self,
-        wet: Grid,
+        loss_fn: LossFn,
```
Comment (jder) [inline src/ocean_emulators/utils/loss.py:135]: Nice, I like the cleanup of using device rather than passing wet around here.

File: src/ocean_emulators/config.py [hunk 8]
```diff
@@ -648,17 +659,69 @@ def build(self, output_dir: Path, device: torch.device) -> Profiler:
...
+def build_loss_fn(
+    loss_cfg: Loss,
+    wet: Grid,
+    y_coord: xr.DataArray,
+    stds: xr.Dataset,
+    device: torch.device,
+) -> LossFn:
+    if isinstance(loss_cfg, str):
+        metric = loss_cfg
+    elif hasattr(loss_cfg, "metric"):
```
Comment (jder) [inline src/ocean_emulators/config.py]: 🐑 this is maybe just too python for me but given we know the loss is either a LossMetric or a DyanmicLossConfig I would probably do something like
```python
match loss_cfg:
  case DynamicLossConfig(metric=metric, limit=limit)
     metric = fn_for_metric(metric)
     return DynamicLoss(…)
  case metric:
     return fn_for_metric(metric)
    ...
```

and then move the later match into fn_for_metric

File: src/ocean_emulators/train.py [hunk 9]
```diff
@@ -232,59 +222,14 @@ def __init__(self, cfg: TrainConfig) -> None:
...
-            case "mse_residual_scaled":
-                logger.info("Using decomposed mse loss with scaled residuals")
-                assert self.data_container.scaling_residuals is not None, (
-                    "With loss of 'mse_residual_scaled' you"
-                    " must supply a scaling_residuals_file"
-                )
-                scale = torch.from_numpy(
-                    (
-                        self.src.stds[self.prognostic_var_names]
-                        / self.data_container.scaling_residuals[
```
Comment (jder) [inline src/ocean_emulators/train.py:261]: Does this mean we can get rid of `scaling_residuals`?

File: src/ocean_emulators/config.py (hunk 5 above)
Comment (jder) [reply src/ocean_emulators/config.py]: Made a suggestion up above. I think a build would also be fine.

File: src/ocean_emulators/config.py [hunk 10]
```diff
@@ -648,17 +659,69 @@ def build(self, output_dir: Path, device: torch.device) -> Profiler:
...
+    limit: bool = Field(
+        description="For dynamic loss, should we limit the range of weighted loss by the variance. Default: off.",
+        default=False,
+    )
+
+
+Loss = LossMetric | DynamicLossConfig
+
+
+def build_loss_fn(
```
Comment (jder) [inline src/ocean_emulators/config.py:659]: 🐑 could move this into utils/loss

File: configs/samudra_om4/train.yaml [hunk 11]
```diff
@@ -8,7 +8,9 @@ epochs: 70
 batch_size: 4
 learning_rate: 0.0002
 scheduler: { type: cosine }
-loss: mse_dynamic_no_limit
+loss:
+  metric: mse
+  type: dynamic
```
Comment (jder) [inline configs/samudra_om4/train.yaml:12]: 🐑 I might put type first since you need to know what type it is to understand the rest of the members

**Review by jder — APPROVED** [review]
Thanks, all optional stuff from me. Looks great!

File: src/ocean_emulators/utils/loss.py (hunk 6 above)
Comment (alxmrs) [reply src/ocean_emulators/utils/loss.py]: Sorry, I think I missed your comment the last time. Thanks for pointing that out (so weird, right?).

File: src/ocean_emulators/config.py (hunk 8 above)
Comment (alxmrs) [reply src/ocean_emulators/config.py]: super slick! I didn't realize that case matching unpacked like that.

File: src/ocean_emulators/train.py (hunk 9 above)
Comment (alxmrs) [reply src/ocean_emulators/train.py:261]: hmm, maybe? I'll double check.

File: src/ocean_emulators/config.py (hunk 10 above)
Comment (alxmrs) [reply src/ocean_emulators/config.py:659]: I switched back and forth on this, but I kind of like that the build methods are all in the config file.

File: configs/samudra_om4/train.yaml (hunk 11 above)
Comment (alxmrs) [reply configs/samudra_om4/train.yaml:12]: Very reasonable.

File: src/ocean_emulators/train.py (hunk 9 above)
Comment (alxmrs) [reply src/ocean_emulators/train.py:261]: Yes! we can! 🎉

File: src/ocean_emulators/utils/loss.py (hunk 6 above)
Comment (jder) [reply src/ocean_emulators/utils/loss.py]: Yes, very weird!
